capture/media-check.js

```diff
@@ -0,0 +1,101 @@
+import { filetypemime } from 'magic-bytes.js'
+
+const TIMEOUT = 2000
+const BYTE_LIMIT = 8192
+
+export function isImageMime (mime) { return typeof mime === 'string' && mime.startsWith('image/') }
+
+export function isVideoMime (mime) { return typeof mime === 'string' && mime.startsWith('video/') }
+
+// adapted from lib/time.js
+function timeoutSignal (timeout) {
+  const controller = new AbortController()
+
+  if (timeout) {
+    setTimeout(() => {
+      controller.abort(new Error(`timeout after ${timeout / 1000}s`))
+    }, timeout)
+  }
+
+  return controller.signal
+}
+
+function hasBasicAuth (res) {
+  const wwwAuth = res.headers.get('www-authenticate') || ''
+  if (res.status === 401 && /basic/i.test(wwwAuth)) return true
+}
+
+async function headMime (url, timeout = TIMEOUT) {
+  const res = await fetch(url, { method: 'HEAD', signal: timeoutSignal(timeout) })
+  // bail on basic authentication requirement
+  if (hasBasicAuth(res)) return null
+
+  return res.headers.get('content-type')
+}
+
+async function readMagicBytes (url, { timeout = TIMEOUT, byteLimit = BYTE_LIMIT } = {}) {
+  const res = await fetch(url, {
+    method: 'GET',
+    // accept image and video, but not other types
+    headers: { Range: `bytes=0-${byteLimit - 1}`, Accept: 'image/*,video/*;q=0.9,*/*;q=0.8' },
+    signal: timeoutSignal(timeout)
+  })
+  // bail on basic authentication requirement
+  if (hasBasicAuth(res)) return { bytes: null, headers: res.headers }
+
+  // stream a small chunk if possible, otherwise read buffer
+  if (res.body?.getReader) {
+    const reader = res.body.getReader()
+    let received = 0
+    const chunks = []
+    try {
+      while (received < byteLimit) {
+        const { done, value } = await reader.read()
+        if (done) break
+        chunks.push(value)
+        received += value.byteLength
+      }
+    } finally {
+      try { reader.releaseLock?.() } catch {}
+      try { res.body?.cancel?.() } catch {}
+    }
+    const buf = new Uint8Array(received)
+    let offset = 0
+    for (const c of chunks) {
+      buf.set(c, offset)
+      offset += c.byteLength
+    }
+    return { bytes: buf, headers: res.headers }
+  } else {
+    const ab = await res.arrayBuffer()
+    const buf = new Uint8Array(ab.slice(0, byteLimit))
+    return { bytes: buf, headers: res.headers }
+  }
+}
+
+export default async function mediaCheck (req, res) {
+  const url = req.params.url
+  if (typeof url !== 'string' || !/^(https?:\/\/)/.test(url)) {
+    return res.status(400).json({ error: 'Invalid URL' })
+  }
+
+  try {
+    // trying with HEAD first, as it's the cheapest option
+    // TODO: should we trust it?
```

Comment:
Outdated TODO?

If not: We can trust that false positives from `HEAD` are negligible, but false negatives are not.

So if HEAD says yes, it's media, we can trust it but not if it says it's not media. That might just be a default response and not according to the spec.